File: Program03/prog3.c
/*
* Alexander Vansteel
* CIS 452
* Project 3
* Whack-a-Mole
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <pthread.h>    // compile with -lpthread
#include <curses.h>     // compile with -lcruses

int pop_dur;            // show duration for moles
int hide_dur;           // hide duration for moles

struct mole_settings {
  int line;             // y position on screen
  int column;           // x position on screen
};

// thread for the moles
void* mole(void *param){
  struct mole_settings *settings = param;
  int line = settings->line;
  int column = settings->column;
  move(column,line);
  printw("a");
  return 0;
}

int main(int argc, char *argv[]){
  int shmid, status;
  key_t key;
  char *shm, *s;
  int grid_width, grid_height, num_moles, max_moles, pop_dur, hide_dur;
  pthread_t *thread;
  struct mole_settings *settings = malloc(sizeof(struct mole_settings));
  char input;
  key = 9876;

  if (argc != 7){
    printf("Incorrect Arguments: ./proj3 grid_width grid_height num_moles max_moles pop_dur hide_dur \n");
    exit(1);
  } else {
    grid_width = atoi(argv[1]);
    grid_height = atoi(argv[2]);
    num_moles = atoi(argv[3]);
    max_moles = atoi(argv[4]);
    pop_dur = atoi(argv[5]);
    hide_dur = atoi(argv[6]);
  }

  thread = malloc(max_moles*sizeof(pthread_t));

  printf("grid_height: %d\n", grid_height);
  // init the game space
  initscr();
  noecho();
  move(grid_height+1,0);
  printw("Welcome to a game of Whack-a-Mole! Press ESC to exit the game.");
  refresh();

  // create game board area and spawn moles into each slot
  int i, j;
  for (i = 0; i < grid_width; i++){
    for (j = 0; j < grid_height; j++){
      settings->column = j;
      settings->line = i;
      if ((status = pthread_create(thread, NULL, mole, (void*)settings)) != 0){
        fprintf(stderr, "thread create error %d: %s\n",
        status, strerror(status));
        exit(1);
      }
    }
  }

  while ((input = getchar()) != 0x1B){
    // TODO: add game logic for main here
    move(grid_height+2,0);
    printw("Character Entered: %c", input);
    refresh();
  }

  endwin();
  return 0;
}

File: Program03/prog3.py
# Alexander Vansteel
# CIS 452
# Whack-a-Mole
# Python version with Multiprocessing

import curses, random, signal, string, sys, time, threading


SIGINT=1


def mole_t(t_num,pop_dur,hide_dur,grid_w,grid_h):
    global semaphore
    global screen
    global char_list
    global coord_list

    while SIGINT:
        time.sleep(random.randrange(0,hide_dur))
        semaphore.acquire()
        char=' '
        while char in char_list: char=random.choice(string.ascii_lowercase)
        y=(random.randrange(0,grid_h)*2)+1
        x=(random.randrange(0,grid_w)*4)+2
        screen.addstr(y,x,char)
        char_list[t_num]=char
        coord_list[t_num]=(y,x)
        time.sleep(random.randrange(pop_dur))
        screen.addstr(y,x,' ')
        char_list[t_num]=' '
        coord_list[t_num]=' '
        semaphore.release()

    time.sleep(6)
    print('Thread '+str(t_num+1)+' ending')


def board_refresh():
    while SIGINT:
        screen.refresh()
        time.sleep(0.01)


def create_board(grid_h, grid_w):
   global screen
   str1 = '+---'
   str2 = '|   '
   top_bottom= ''
   side = ''

   for w in range(grid_w):
       top_bottom += str1
       side += str2

   top_bottom += '+'
   side += '|'

   for h in range(grid_h):
       screen.addstr(h*2, 0, top_bottom)
       screen.addstr((h*2)+1, 0, side)
       screen.addstr((h*2)+2, 0, top_bottom)


def main(stdscr,grid_w,grid_h,moles,max_moles,pop_dur,hide_dur):
    global SIGINT

    global screen
    screen=stdscr

    global semaphore
    semaphore=threading.BoundedSemaphore(max_moles)

    global key
    key=' '

    screen.clear()
    screen.addstr(2*grid_h+2,0,'Welcome to Whack-a-Mole! (Press esc to exit)')
    screen.refresh()
    curses.curs_set(0)

    create_board(grid_h,grid_w)

    global char_list
    char_list=[' ']*moles
    global coord_list
    coord_list=[' ']*moles

    try:
        for t_num in range(moles):
            t=threading.Thread(target=mole_t,args=(t_num,pop_dur,hide_dur,grid_w,grid_h,))
            t.start()
    except:
        raise Exception("Error: Failed to start threads.")

    t=threading.Thread(target=board_refresh,args=())
    t.start()

    score=0
    misses=0
    while key != 27: # esc char
        key=screen.getch()
        screen.addstr(2*grid_h+3,0,'Key Pressed: '+chr(key))

        if chr(key) in char_list:
            index=char_list.index(chr(key))
            score=score+1
            screen.addstr(2*grid_h+4,0,'Score: '+str(score))
            screen.addstr(int(coord_list[index][0]),int(coord_list[index][1]),' ')
            char_list[index]=' '
            coord_list[index]=' '
        else:
            misses+=1
            screen.addstr(2*grid_h+5,0,'Misses: '+str(misses))

        if misses > 9:
            SIGINT=0
            curses.flash()
            curses.flushinp()
            screen.clear()
            screen.refresh()
            screen.addstr(0,0,'Oh no! You have missed the mole too many times and lost the game...',curses.A_BLINK)
            screen.addstr(1,0,'Closing game in 5 sec...',curses.A_BLINK)
            screen.refresh()
            time.sleep(5)
            key=27

    SIGINT=0
    curses.curs_set(2)


if __name__=='__main__':
    grid_w=int(sys.argv[1])
    grid_h=int(sys.argv[2])
    moles=int(sys.argv[3])
    max_moles=int(sys.argv[4])
    pop_dur=int(sys.argv[5])
    hide_dur=int(sys.argv[6])

    if (len(sys.argv) < 7):
        print('Incorrect number of arguments')
        print('Usage: '+sys.argv[0]+' [width] [height] [number of moles] [max moles] [pop duration(ms)] [hide duration(ms)]')
        sys.exit()

    try:
        curses.wrapper(main,grid_w,grid_h,moles,max_moles,pop_dur,hide_dur)
    except KeyboardInterrupt:
        print('Received KeyboardInterrupt Exception: Exiting...')
        sys.exit()
